test/FullRange.t.sol

```diff
@@ -0,0 +1,453 @@
+
+        (bool owed, address liquidityToken) = fullRange.poolInfo(id);
+
+        assertEq(MockERC20(token0).balanceOf(address(this)), prevBalance0 - 25 ether);
+        assertEq(MockERC20(token1).balanceOf(address(this)), prevBalance1 - 25 ether);
+
+        assertEq(UniswapV4ERC20(liquidityToken).balanceOf(address(this)), 25 ether + 1 - 1000);
+        assertEq(owed, false);
+    }
+
+    function testSwapAddLiquiditySucceeds() public {
+        manager.initialize(key, SQRT_RATIO_1_1);
+
+        uint256 prevBalance0 = MockERC20(token0).balanceOf(address(this));
+        uint256 prevBalance1 = MockERC20(token1).balanceOf(address(this));
+        (, address liquidityToken) = fullRange.poolInfo(id);
+
+        fullRange.addLiquidity(address(token0), address(token1), 3000, 10 ether, 10 ether, address(this), MAX_DEADLINE);
+
+        assertEq(UniswapV4ERC20(liquidityToken).balanceOf(address(this)), 10 ether - 1000);
+        assertEq(MockERC20(token0).balanceOf(address(this)), prevBalance0 - 10 ether);
+        assertEq(MockERC20(token0).balanceOf(address(this)), prevBalance1 - 10 ether);
+
+        vm.expectEmit(true, true, true, true);
+        emit Swap(
+            id, address(swapRouter), 1 ether, -906610893880149131, 72045250990510446115798809072, 10 ether, -1901, 3000
+        );
+
+        IPoolManager.SwapParams memory params =
+            IPoolManager.SwapParams({zeroForOne: true, amountSpecified: 1 ether, sqrtPriceLimitX96: SQRT_RATIO_1_2});
+        PoolSwapTest.TestSettings memory settings =
+            PoolSwapTest.TestSettings({withdrawTokens: true, settleUsingTransfer: true});
+
+        snapStart("swap");
+        swapRouter.swap(key, params, settings);
+        snapEnd();
+
+        (bool owed,) = fullRange.poolInfo(id);
+
+        assertEq(MockERC20(token0).balanceOf(address(this)), prevBalance0 - 10 ether - 1 ether);
+        assertEq(MockERC20(token1).balanceOf(address(this)), prevBalance1 - 9093389106119850869);
+
+        assertEq(owed, true);
+
+        fullRange.addLiquidity(address(token0), address(token1), 3000, 5 ether, 5 ether, address(this), MAX_DEADLINE);
+
+        (owed,) = fullRange.poolInfo(id);
+
+        assertEq(UniswapV4ERC20(liquidityToken).balanceOf(address(this)), 14546694553059925434 - 1000);
+        assertEq(owed, true);
+    }
+
+    function testTwoSwaps() public {
+        manager.initialize(key, SQRT_RATIO_1_1);
+
+        fullRange.addLiquidity(address(token0), address(token1), 3000, 10 ether, 10 ether, address(this), MAX_DEADLINE);
+
+        IPoolManager.SwapParams memory params =
+            IPoolManager.SwapParams({zeroForOne: true, amountSpecified: 1 ether, sqrtPriceLimitX96: SQRT_RATIO_1_2});
+        PoolSwapTest.TestSettings memory settings =
+            PoolSwapTest.TestSettings({withdrawTokens: true, settleUsingTransfer: true});
+
+        snapStart("swap first");
+        swapRouter.swap(key, params, settings);
```

Comment:
once again, a storage reference for gas... would go down the file and cache in memory when taking gas estimates!